### ISSUE TITLE
Doc2Vec instead of Word2Vec model for Gensim featurizer?

### ISSUE THREAD

**Issue by predoctech** [RasaHQ/rasa-nlu-examples#52]
Wondering if doc2vec, instead of word2vec, model can be trained and deployed to Gensim featurizer?

Also if HFTransformer is chosen as the Language Model, thus the corresponding tokenizer and featurizer are used in NLU pipeline, can the Gensim featurizer can still be added to the pipeline to improve domain specific processing?

Thanks.

**Comment by sara-tagger** [comment]
Thanks for the issue, @tttthomasssss will get back to you about it soon!

###### You may find help in the [docs](https://rasa.com/docs/) and the [forum,](https://forum.rasa.com/) too 🤗

**Comment by koaning** [comment]
The featurizers here just take the text of the utterance/token and featurize from there. I think the doc2vec from gensim should be fine, assuming they use the same `KeyedVectors`-api.

Note that the featurizers here are built such that pretrained embedding can be applied in Rasa. What we host here won't do the training for you, you need to do that yourself beforehand either by training manually or by downloading.

**Comment by predoctech** [comment]
Thanks @koaning.  I'm new to Rasa so I may not fully comprehend what you've explained.  Let me restate from a user perspective to see if I've missed anything.
The reason I asked about using a doc2vec, instead of a word2vec, model is that I wish to alter the behavior of pretrained featurizers (ie. LanguageModelFeaturizer) by utilizing custom-trained embeddings.  It seems to me word2vec will put emphasis on custom words but if the idea is to put emphasis on custom phrases then should doc2vec be used instead?
Or should we first clarify whether that's the purpose of utilizing GensimFeaturizer in the first place?

I understand the Gensim model will need to be trained with custom data first.  The question is really once that ".kv" file is produced, how its inclusion in the NLU pipeline will impact on existing Featurizer?  And whether such an impact can be at the phrase instead of word level?

Please help me to understand, much appreciated.

**Comment by koaning** [comment]
Currently the GensimFeaturizer is able to add a numerical vector to each word but also to the entire text. Here's how that works;

![image](https://user-images.githubusercontent.com/1019791/98347263-7ab5bf00-2017-11eb-823a-0896c6cb0caf.png)

All of the vectors that are drawn will be added to any set of features that already exist in the pipeline. That means that the intent prediction will also be influenced by any pretrained embedding that have been added.

**Details** 

Now the difference between word2vec and doc2vec (as far as I can tell from reading the [paper](https://cs.stanford.edu/~quocle/paragraph_vector.pdf)) is that word2vec tries to make embedding for words while doc2vec is trying to add context by adding paragraph features. It works via a similar method but there's a trick they do with a sliding window on a paragraph matrix to try to generalise the sequential information in a document. Because of this there are some differences in the training procedure. Looking at the documentation of gensim [here](https://radimrehurek.com/gensim/models/doc2vec.html) I also get the impression that it is using a slightly different API which means that the doc2vec approach is currently not supported. That said, the standard word2vec should still allow you to train embeddings for your domain.

**Comment by koaning** [comment]
**Final Comment**

I've only checked this for English but it seems sofar that when you're starting out ... you shouldn't worry about embeddings too much just yet. The problem, especially when you're just starting out, is most likely that you do not have enough labels/user feedback yet to really get a good idea of what you should be optimising towards.

![image](https://user-images.githubusercontent.com/1019791/98349429-4394dd00-201a-11eb-9f92-aa4ad2181e64.png)

I typically start thinking about embeddings only after I've done a few rounds of user feedback.
